### PR TITLE
Update Agent interface and improve documentation

### LINKED ISSUE #1862: As a user of the python controller/agent infrastructure I would like to be able to call Controller.run() multiple times
> estimate 1

At least one additional method for the Agent class must be defined to tell the agent object that a new run is beginning.  May also want another method that defines the end of a run for the agent.  The Agent.get_report() method could be used for this purpose, but seems to be overloading the interface.

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Enables an Agent object to be used with a single controller to
  execute multiple run() calls.
- Fixes #1862 from github issues

### CODE REVIEW

File: service/geopmdpy/runtime.py [hunk 1]
```diff
@@ -265,9 +317,10 @@ def run(self):
         """
         try:
             pid = subprocess.Popen(self._argv)
+            self._agent.run_begin(policy)
             for loop_idx in TimedLoop(self._update_period, self._num_update):
                 if pid.poll() is not None:
-                    sys.stderr.write('Ths app process has ended.  return code = {}'.format(pid.returncode))
+                    sys.stderr.write('The app process has ended.  return code = {}'.format(pid.returncode))
```
Comment (fkeceli) [inline service/geopmdpy/runtime.py]: Can you put a \n at the end of the string?

**Review by fkeceli — COMMENTED** [review]
Few things:

- What about moving argv to run() method as I brought up before?
- In the Controller __init__ (...timeout) documentation it is mentioned that "the agent will run indefinitely (until killed by another process) if timeout == 0. I know that run will end if the app finishes naturally (not killed). I think that doc can be clearer.

**Comment by cmcantalupo** [comment]
Thanks for the feedback @fkeceli.  I think the last two patches address your comments.

**Review by fkeceli — DISMISSED** [review]
Good changes. Thanks Chris. I actually hit "Do not throw on calls to ServiceIOGroup::read_batch()/write_batch()" bug in Saruman.Thanks for making the change.

File: service/geopmdpy/runtime.py [hunk 2]
```diff
@@ -181,7 +227,7 @@ def update(self, signals):
         raise NotImplementedError('Agent is an abstract base class')
 
     def get_period(self):
-        """Get the target time interal for the control loop
+        """Get the target time internal for the control loop
```
Comment (bgeltz) [inline service/geopmdpy/runtime.py]:
```suggestion
        """Get the internal target time for the control loop
```
or just
```suggestion
        """Get the target time for the control loop
```

It's a getter.  It's assumed it's returning internal state.

File: service/integration/test/test_controller.py [hunk 3]
```diff
@@ -69,24 +83,29 @@ def get_report(self):
...
         sys.stderr.write(help_msg)
         return err
+    cpu_frequency_max = float(sys.argv[1])
     agent = LocalAgent()
-    controller = geopmdpy.runtime.Controller(agent, sys.argv[1:])
-    print(controller.run())
-    return err
+    controller = geopmdpy.runtime.Controller(agent)
+    print(controller.run(sys.argv[2:], cpu_frequency_max))
+    return controller.returncode()
```
Comment (bgeltz) [inline service/integration/test/test_controller.py:108]: Nice.

File: service/src/ServiceIOGroup.cpp [hunk 4]
```diff
@@ -171,14 +171,12 @@ namespace geopm
 
     void ServiceIOGroup::read_batch(void)
     {
-        throw Exception("ServiceIOGroup::read_batch()",
-                        GEOPM_ERROR_NOT_IMPLEMENTED, __FILE__, __LINE__);
+
```
Comment (bgeltz) [inline service/src/ServiceIOGroup.cpp]: Maybe at least a comment here saying these are still unimplemented, and not intentionally empty?

**Review by bgeltz — CHANGES_REQUESTED** [review]

File: service/geopmdpy/runtime.py (hunk 2 above)
Comment (cmcantalupo) [reply service/geopmdpy/runtime.py]: I think that was supposed to be *interval* not internal

**Review by bgeltz — APPROVED** [review]

**Pull request merged**